go/parquet/internal/encoding/byte_stream_split.go

```diff
@@ -0,0 +1,296 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encoding
+
+import (
+	"math"
+
+	"github.com/apache/arrow/go/v17/arrow"
+	"github.com/apache/arrow/go/v17/internal/bitutils"
+	"github.com/apache/arrow/go/v17/internal/utils"
+	"github.com/apache/arrow/go/v17/parquet"
+	"golang.org/x/xerrors"
+)
+
+type NumericByteStreamSplitType interface {
+	float32 | float64 | int32 | int64
+}
+
+type ByteStreamSplitType interface {
+	NumericByteStreamSplitType | parquet.FixedLenByteArray
+}
+
+// putByteStreamSplitNumeric is a generic implementation of the BYTE_STREAM_SPLIT encoding for fixed-width numeric types.
+func putByteStreamSplitNumeric[T NumericByteStreamSplitType](in []T, enc TypedEncoder, sink *PooledBufferWriter) {
```

Comment:
would this crack the previously put bytes within `sink`?

Reply:
The last line of this function is `sink.pos += bytesNeeded`, which ensures subsequent bytes are written _after_ the chunk written by the current call. Is that the issue you were referring to?

Reply:
I think we need to build up the full buffer of values and only do the encoding during the call to flush when we have all the values we're going to write for a given data page. Otherwise this won't decode properly.

Reply:
> which ensures subsequent bytes are written after the chunk written by the current call. Is that the issue you were referring to?

Assume we have 2 value and per value has 4 streams. When put the first value, the 0-3 bytes is written. When put the second value, it "should" use "1, 3, 5, 7" byte